File: dynamic_programming/houseRobber.cc
// Time: O(n)
// Space: O(n)
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.size() == 1) {
            // To avoid out of bounds error from setting base case
            return nums[0];
        }
        
        int n = int(nums.size());
        vector<int> dp(n);

        // Base cases
        dp[0] = nums[0];
        dp[1] = max(nums[0], nums[1]);
        
        for (int i = 2; i < n; i++) {
            // Recurrence relation
            dp[i] = max(dp[i - 1], dp[i - 2] + nums[i]);
        }
        
        return dp[n - 1];
    }
};

File: dynamic_programming/houseRobber.py
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        """
        Algorithm intuition:
            - maximize money robbed -> dp
            - state variable -> i (money robbed)
        Time: O(n)
        Space: O(1)
        """
        n = len(nums)

        if n == 1:
            return nums[0]

        backTwo = nums[0]
        backOne = max(nums[0], nums[1])

        for i in range(2, n):
            backOne, backTwo = max(backOne, backTwo + nums[i]), backOne

        return backOne
